torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp

```diff
@@ -0,0 +1,917 @@
+  group->consumer_edges.clear();
+
+  return removed_edges;
+}
+
+void SegmentCandidateFinder::mergeNodes() {
+  while (!to_merge_.empty()) {
+    auto group1 = *to_merge_.begin();
+    auto group2 = group1->merge_with;
+    to_merge_.erase(group1);
+    to_merge_.erase(group2);
+
+    clean_up_groups_.emplace(group1);
+    clean_up_groups_.emplace(group2);
+
+    // Make the new joined node
+    auto joined_group = segmented_fusion_->newGroup();
+
+    joined_group->input_vals =
+        uniqueValConcat({group1->input_vals, group2->input_vals});
+
+    joined_group->output_vals =
+        uniqueValConcat({group1->output_vals, group2->output_vals});
+
+    joined_group->exprs_ = group1->exprs_;
+    joined_group->exprs_.insert(
+        joined_group->exprs_.end(),
+        group2->exprs_.begin(),
+        group2->exprs_.end());
+
+    auto producer_edges = getMergedProducerEdges(group1, group2);
+    // Connect joined group to resulting neighbors
+    for (auto edge : producer_edges) {
+      auto from = edge->from;
+      auto val = edge->val;
+
+      auto new_edge = segmented_fusion_->newEdge(from, joined_group, val);
+      joined_group->producer_edges.push_back(new_edge);
+      from->consumer_edges.push_back(new_edge);
+    }
+
+    auto consumer_edges = getMergedConsumerEdges(group1, group2);
+
+    for (auto edge : consumer_edges) {
+      auto to = edge->to;
+      auto val = edge->val;
+
+      auto new_edge = segmented_fusion_->newEdge(joined_group, to, val);
+      joined_group->consumer_edges.push_back(new_edge);
+      edge->to->producer_edges.push_back(new_edge);
+    }
+
+    joined_group->setHeuristic(deriveHeuristic(joined_group));
+  }
+
+  for (auto group : clean_up_groups_) {
+    auto disconnected_edges = disconnectGroup(group);
+    clean_up_edges_.insert(
+        disconnected_edges.begin(), disconnected_edges.end());
+  }
+
+  edges().erase(
+      std::remove_if(
+          edges().begin(),
+          edges().end(),
+          [this](SegmentedEdge* edge) {
+            if (this->clean_up_edges_.find(edge) !=
+                this->clean_up_edges_.end()) {
+              return true;
+            };
+            return false;
+          }),
+      edges().end());
+
+  groups().erase(
+      std::remove_if(
+          groups().begin(),
+          groups().end(),
+          [this](SegmentedGroup* group) {
+            if (this->clean_up_groups_.find(group) !=
+                this->clean_up_groups_.end()) {
+              return true;
+            };
+            return false;
+          }),
+      groups().end());
+
+  clean_up_edges_.clear();
+  clean_up_groups_.clear();
+}
+
+namespace {
+
+// Guard to temporarily change the inputs and outputs of a fusion. On
+// destruction will return fusion to original state.
+// Not used temporarily but will be useful when adding more mergin heuristics
+class FusionSegmentGuard {
+ public:
+  FusionSegmentGuard() = delete;
+  FusionSegmentGuard(FusionSegmentGuard const&) = delete;
```

Comment:
`class FusionSegmentGuard : public NonCopyable` instead of manually deleting the copy operations

Reply:
Updated, Thanks!